File: Python/stage_II.py
# Andoni Garcia. 2014.
# My source code for Stage I: Needle in a Haystack

from urllib.request import Request, urlopen
from json import dumps, loads

token = "yicZ5DfAT6"
response = None
index = None

def GetHaystack():
	global token, response
	url = "http://challenge.code2040.org/api/haystack"
	values = {'token':token}
	JSONData = dumps(values).encode('utf-8')
	req = Request(url, data = JSONData, headers = {'content-type':'application/json'})
	resp = urlopen(req)
	response = loads(resp.read().decode('utf-8'))['result']
	print("Your needle is: " + response['needle'])
	print("Your haystack is: " + str(response['haystack']))

def FindIndex1():
	global response, index
	try:
		index = response['haystack'].index(response['needle'])
	except ValueError as e:
		print("Error: Needle not in haystack")
		print(e.code())
	else:
		print("Needle is at position: " + str(index))

# Again, slower but uses loops (the point of the exercise)
def FindIndex2():
	global response, index
	needle = response['needle']
	haystack = response['haystack']
	for i in range(len(haystack)):
		if needle == haystack[i]:
			index = i
			print("Needle is at position: " + str(index))
			return
	print("Error: Needle is not in the haystack")

def PostIndex():
	global token, index
	url = "http://challenge.code2040.org/api/validateneedle"
	values = {'token':token,
			  'needle':index}
	JSONData = dumps(values).encode('utf-8')
	req = Request(url, data = JSONData, headers = {'content-type':'application/json'})
	resp = urlopen(req)
	result = loads(resp.read().decode('utf-8'))['result']
	print("Result: " + result)

def StageII():
	GetHaystack()
	FindIndex1()
	PostIndex()

StageII()

File: JavaScript/stage_II.js
// Andoni Garcia. 2014.
// My source code for Stage II: Needle in a Haystack

var token = "yicZ5DfAT6"; 
var xmlhttp = new XMLHttpRequest();
var response;
var index;


function GetHaystack(){
	var url = "http://challenge.code2040.org/api/haystack";
	var values = {"token":token};
	var JSONData = JSON.stringify(values);
	xmlhttp.onreadystatechange = function(){
		if (xmlhttp.readyState == 4 && xmlhttp.status == 200){
			response = JSON.parse(xmlhttp.responseText).result;
			console.log("Your needle is: " + response.needle);
			console.log("Your haystack is: " + response.haystack);
		}
	}
	xmlhttp.open("POST", url, false);
	xmlhttp.setRequestHeader("content-type", "application/json");
	xmlhttp.send(JSONData);
}

function FindIndex1(){
	var needle = response.needle;
	var haystack = response.haystack;
	index = haystack.indexOf(needle);
	console.log("The needle is at index: " + index);
}

function FindIndex2(){
	var needle = response.needle;
	var haystack = response.haystack;
	for (i = 0; i < haystack.length; i++){
		if (needle === haystack[i]){
			index = i;
			console.log("The needle is at index: " + index);
			return;
		}
	}
}

function PostIndex(){
	var url = "http://challenge.code2040.org/api/validateneedle";
	var values = {"token":token,
				  "needle":index};
	var JSONData = JSON.stringify(values);
	xmlhttp.onreadystatechange = function(){
		if (xmlhttp.readyState == 4 && xmlhttp.status == 200){
			var result = JSON.parse(xmlhttp.responseText);
			console.log("Result: " + result.result);
		}
	}
	xmlhttp.open("POST", url, false);
	xmlhttp.setRequestHeader("content-type", "application/json");
	xmlhttp.send(JSONData);
}

// HIGHLY ADVISED TO COMMENT OUT PRINT STATEMENTS BEFORE EXECUTING TIMEFUNCS
function TimeFuncs(){
	GetHaystack();
	console.time('FindIndex1');
	FindIndex1();
	console.timeEnd('FindIndex1');
	console.time('FindIndex2');
	FindIndex2();
	console.timeEnd('FindIndex2');
}


function StageII(){
	GetHaystack();
	FindIndex1();
	PostIndex();
}

StageII();